apm-agent-plugins/apm-java-concurrent-plugin/src/test/java/co/elastic/apm/agent/concurrent/AsyncTraceMethodInstrumentationTest.java

```diff
@@ -0,0 +1,149 @@
+import co.elastic.apm.agent.bci.ElasticApmAgent;
+import co.elastic.apm.agent.bci.methodmatching.MethodMatcher;
+import co.elastic.apm.agent.configuration.CoreConfiguration;
+import co.elastic.apm.agent.configuration.SpyConfiguration;
+import co.elastic.apm.agent.configuration.converter.TimeDuration;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.ElasticApmTracerBuilder;
+import net.bytebuddy.agent.ByteBuddyAgent;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInfo;
+import org.stagemonitor.configuration.ConfigurationRegistry;
+
+import java.util.Arrays;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+class AsyncTraceMethodInstrumentationTest {
+
+    private MockReporter reporter;
+    private ElasticApmTracer tracer;
+    private CoreConfiguration coreConfiguration;
+
+    @BeforeEach
+    void setUp(TestInfo testInfo) {
+        reporter = new MockReporter();
+        ConfigurationRegistry config = SpyConfiguration.createSpyConfig();
+        coreConfiguration = config.getConfig(CoreConfiguration.class);
+        when(coreConfiguration.getTraceMethods()).thenReturn(Arrays.asList(
+            MethodMatcher.of("private co.elastic.apm.agent.concurrent.AsyncTraceMethodInstrumentationTest$TestAsyncTraceMethodsClass#*"))
+        );
+
+        Set<String> tags = testInfo.getTags();
+        if (!tags.isEmpty()) {
+            when(coreConfiguration.getTraceMethodsDurationThreshold()).thenReturn(TimeDuration.of(tags.iterator().next()));
+        }
+
+        tracer = new ElasticApmTracerBuilder()
+            .configurationRegistry(config)
+            .reporter(reporter)
+            .build();
+        ElasticApmAgent.initInstrumentation(tracer, ByteBuddyAgent.install());
+    }
+
+    @AfterEach
+    void tearDown() {
+        ElasticApmAgent.reset();
+    }
+
+    @Test
+    void testWithDefaultConfig() {
+        new TestAsyncTraceMethodsClass().invokeAsync();
+        assertThat(reporter.getTransactions()).hasSize(1);
+        assertThat(reporter.getSpans()).hasSize(4);
+    }
+
+    @Test
+    @Tag("200ms")
+    void testWithHighThreshold() {
+        new TestAsyncTraceMethodsClass().invokeAsync();
+        assertThat(reporter.getTransactions()).hasSize(1);
+        assertThat(reporter.getSpans()).hasSize(2);
+    }
+
+    @Test
+    @Tag("50ms")
+    void testWithCrossedThreshold() {
+        new TestAsyncTraceMethodsClass().invokeAsync();
+        assertThat(reporter.getTransactions()).hasSize(1);
+        assertThat(reporter.getSpans()).hasSize(4);
+    }
+
+
+    private static class TestAsyncTraceMethodsClass {
+
+        /**
+         * Calling this method results in this method call tree:
```

Comment:
seems like the formatting got messed up, maybe your local history still contains the old version